spec/sec-string-objects-patch.html

```diff
@@ -1,39 +1,74 @@
     </emu-note>
   </emu-clause>
   <emu-clause id="sec-string.prototype.matchall">
-    <h1>String.prototype.matchAll ( _regexp_ [, _mapfn_ ] )</h1>
+    <h1>String.prototype.matchAll ( _regexp_<ins> [, _options_ ]</ins> )</h1>
     <p>Performs a regular expression match of the String representing the *this* value against _regexp_ and returns an iterator. Each iteration result’s value is an Array object containing the results of the match, or *null* if the String did not match.</p>
     <p>When the `matchAll` method is called, the following steps are taken:</p>
     <emu-alg>
     1. Let _O_ be ? RequireObjectCoercible(*this* value).
-    1. <ins>If _mapfn_ is not *undefined* and IsCallable(_mapfn_) is *false*, throw a *TypeError* exception.</ins>
+    1. <ins>If _options_ is not present, let _options be *undefined*.</ins>
+    1. <ins>Let _matchOptions_ be ? ToMatchOptions(_options_).</ins>
     1. If _regexp_ is neither *undefined* nor *null*, then
       1. Let _matcher_ be ? GetMethod(_regexp_, @@matchAll).
       1. If _matcher_ is not *undefined*, then
-      1. Return ? Call(_matcher_, _regexp_, &laquo; _O_<ins>, _mapfn_</ins> &raquo;).
+        1. Return ? Call(_matcher_, _regexp_, &laquo; _O_<ins>, ! FromMatchOptions(_matchOptions_)</ins> &raquo;).
     1. Let _S_ be ? ToString(_O_).
     1. Let _rx_ be ? RegExpCreate(_regexp_, `"g"`).
-    1. Return ? Invoke(_rx_, @@matchAll, &laquo; _S_<ins>, _mapfn_</ins> &raquo;).
+    1. Return ? Invoke(_rx_, @@matchAll, &laquo; _S_<ins>, ! FromMatchOptions(_matchOptions_)</ins> &raquo;).
     </emu-alg>
     <emu-note>The `matchAll` function is intentionally generic, it does not require that its *this* value be a String object. Therefore, it can be transferred to other kinds of objects for use as a method.</emu-note>
     <emu-note>Similarly to `String.prototype.split`, `String.prototype.matchAll` is designed to typically act without mutating its inputs.</emu-note>
   </emu-clause>
+  <emu-clause id="sec-string.prototype.replace">
+    <h1>String.prototype.replace ( _searchValue_, _replaceValue_<ins>[, _options_ ]</ins> )</h1>
+    <p>When the `replace` method is called with arguments _searchValue_ and _replaceValue_ <ins>and optional argument _options_</ins>, the following steps are taken:</p>
+    <emu-alg>
+    1. Let _O_ be ? RequireObjectCoercible(*this* value).
+    1. <ins>If _options_ is not present, let _options be *undefined*.</ins>
+    1. <ins>Let _matchOptions_ be ? ToMatchOptions(_options_).</ins>
+    1. If _searchValue_ is neither *undefined* nor *null*, then
+      1. Let _replacer_ be ? GetMethod(_searchValue_, @@replace).
+      1. If _replacer_ is not *undefined*, then
+        1. Return ? Call(_replacer_, _searchValue_, &laquo; _O_, _replaceValue_<ins>, ! FromMatchOptions(_matchOptions_)</ins> &raquo;).
+    1. <ins>If _matchOptions_.[[Captures]] is not `"strings"`, throw a *RangeError* exception.</ins>
```

Comment:
this seems like something that should be thrown higher up and unconditionally, right after `options` is first processed. Can this be moved up, or even better, maybe `ToMatchOptions` could take an arg that has it do this check?

Reply:
This check is only necessary for `"foo".replace("bar", "baz", { captures: "indices" })`, since indices wouldn't make sense in this case.

Reply:
Putting it earlier would just mean duplicating steps 3.1 and 3.2.